[magnum] Add recipe

jgsogo opened the pull request:
- [x] https://github.com/conan-io/conan-center-index/pull/6290
- [x] https://github.com/conan-io/conan-center-index/pull/6942
- [x] https://github.com/conan-io/conan-center-index/pull/6387
- [x] https://github.com/conan-io/conan-center-index/pull/6388

SSE4 commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,389 @@
+class MagnumConan(ConanFile):
+    name = "magnum"
+    description = "Lightweight and modular C++11/C++14 graphics middleware for games and data visualization"
+    license = "MIT"
```
```suggestion
    license = "MIT"
    short_paths = True
```

jgsogo commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+        "target_gl": True,
+        "target_gles": True, # TODO: Here we probably have a CHOICE OPTION
+        "target_gles2": True, # TODO: Here we probably have a CHOICE OPTION
+        "target_desktop_gles": True, # TODO: Here we probably have a CHOICE OPTION
```
Pending. I think it is better to offer an option like:

```
target_gles: ["gles2", "gles3", "desktop_gles"],
```

wdyt?

jgsogo commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+                              "egl_system_FOUND")
+        tools.replace_in_file(os.path.join(self._source_subfolder, "src", "Magnum", "Platform", "CMakeLists.txt"),
+                              "EGL::EGL",
+                              "egl::egl")
```
I need to understand why we are renaming to `egl_system` and, if it would be better to consume the `FindEGL.cmake` file provided by the library or CMake one.

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+    description = "Lightweight and modular C++11/C++14 graphics middleware for games and data visualization"
+    license = "MIT"
+    short_paths = True
+    topics = ("conan", "corrade", "graphics", "rendering", "3d", "2d", "opengl")
```
```suggestion
    topics = ("magnum", "graphics", "middleware", "graphics", "rendering", "gamedev", "opengl", "3d", "3d", "opengl", "game-engine")
```

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+        "fPIC": True,
+        "shared_plugins": True,
+
+        "target_gl": True,
```
It looks like the names of the options are 1:1 copies of the cmake script.
I would rename these options such that they can be understood without reading the cmake script.

`target_gl` -> `enable_gl`
target_gles`, `target_gles` and `target_desktop_gles` can all be enabled (they are not not completely exclusive -> there exists a dependence). I would rename them to `enable_*`.

jgsogo replied:
About options, yes, they are 1-1 copies of CMake script. I really needed it while developing the recipe and figuring out all the logic. It was a bit hard sometimes :sweat:  I agree we can change them, letś try to choose the best names for them.

The project has options `with_gl` that enables GL component an requires one more dependency, but also `target_gl` that will add OpenGL support to other libraries (adds some extra sources). It is the same for options `with/target_vk`.

jgsogo replied:
About GL, I would say those options (gles, gles2 and desktop) are mutually exclusive and all require `with_gl`. This is from sources:

```cmake
if(WITH_GL)
    if(NOT TARGET_GLES OR TARGET_DESKTOP_GLES)
        # OpenGL library preference. Prefer to use GLVND, since that's the
        # better approach nowadays, but allow the users to override it from
        # outside in case it is broken for some reason (Nvidia drivers in
        # Debian's testing (Buster) -- reported on 2019-04-09).
        if(NOT CMAKE_VERSION VERSION_LESS 3.10 AND NOT OpenGL_GL_PREFERENCE)
            set(OpenGL_GL_PREFERENCE GLVND)
        endif()
        find_package(OpenGL REQUIRED)
    elseif(TARGET_GLES2)
        find_package(OpenGLES2 REQUIRED)
    else()
        find_package(OpenGLES3 REQUIRED)
    endif()
else()
    if(TARGET_GL)
        message(FATAL_ERROR "WITH_GL is OFF but TARGET_GL is enabled. Hmm.")
    endif()
    ...
```

jgsogo replied:
I think I will rename `with_xxxx` to `xxxx` (it says about the libraries that are built), but I feel like that the `target_xxx` should remain the same (they say about the _supported_ features in existing libraries)

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+        "target_headless": True,
+        "target_vk": True,
+
+        "with_audio": False,
```
We mostly use the `with_` prefix when an options causes an external dependency to be pulled in.
I would rename these to `audio` or `enable_audio`.

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+        "with_trade": True,
+        "with_vk": True,
+
+        "with_androidapplication": True,
```
These option look unreadable to me.
How about renaming these to: `enable_app_android`.
I believe the extra `_` makes it easier for our brains to parse an option.

jgsogo replied:
Absolutely. Now that I have stabilized the recipe I can afford changing the options names :smile:

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+
+        # Audio
+        if self.options.with_audio:
+            raise Exception("Component not created")
```
Is this a FIXME?

jgsogo replied:
I haven't build the library with this option activated. It requires `OpenAL` and I didn't realize it was already available. Letś see if it works.

And yes, this is like a TODO (same for other libraries for iOS, emsdk or Android): if someone manages to build them, they will need to populate the required component.

madebr replied:
Can you please output a nice message + explicit `FIXME` comment then?

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+        if self.options.with_gl:
+            self.requires("opengl/system")
+        if self.options.with_vk:
+            self.requires("vulkan-loader/1.2.182")
```
```suggestion
            self.requires("vulkan-loader/1.2.190")
```

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,710 @@
+            self.requires("sdl/2.0.16")
+
+    def build_requirements(self):
+        self.build_requires("corrade/{}".format(self.version))
```
Shouldn't this be added conditionally when using 2 profile build?
I had to do this in the `autoconf` recipe:
```suggestion
        if hasattr(self, "settings_build"):
            self.build_requires("corrade/{}".format(self.version))
```

jgsogo replied:
There is no reason to do it, specially when the versions of `magnum` and `corrade` (both the requires and the build-requires) will probably be always the same. It is very unlikely that someone will override the `corrade` requirement. I really prefer to keep the recipe simpler.

jgsogo commented:
@madebr , now we probably have something to review. I've renamed the options (no more `with_`, use snake_case), enabled audio components, fixed some typos,...

madebr commented on recipes/magnum/all/test_package/test_package.cpp:
```diff
@@ -0,0 +1,20 @@
+/*
+    I would like to use some windowless application to test, like
+    https://github.com/mosra/magnum-bootstrap/tree/windowless
+    but it doesn't work in CI.
```
Some more information would be useful.
e.g. that `EGL_NOT_INITIALIZED` is returned on CI
https://github.com/conan-io/conan-center-index/pull/7095#issuecomment-913841567

madebr commented on recipes/magnum/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
```
```suggestion
project(test_package CXX)
```

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,723 @@
+        if self.settings.os == "Linux":
+            del self.options.cgl_context
+            del self.options.windowless_cgl_application
+            del self.options.wgl_context
```
I think `wgl_context` (and every option containing `wgl`) is Windows-only:
https://github.com/mosra/magnum/blob/03b7adf71b1b50d2d1d2064b4a185355ee19dd5f/CMakeLists.txt#L167

jgsogo replied:
I can build `wgl_context` in Macos. TBH, I have no idea if it is usable or not. I disabled only the options that were failing for each OS, I find it very hard to reproduce the logic encoded in the build system.

madebr commented on recipes/magnum/all/conanfile.py:
```diff
@@ -0,0 +1,723 @@
+        "glx_application": True,
+        "sdl2_application": True,
+        "xegl_application": True,
+        "windowless_cgl_application": True,
+        "windowless_egl_application": True,
+        "windowless_glx_application": True,
+        "windowless_ios_application": True,
+        "windowless_wgl_application": True,
+        "windowless_windows_egl_application": True,
+
+        "cgl_context": True,
+        "egl_context": True,
+        "glx_context": True,
+        "wgl_context": True,
```
Do we care about some illegal option combinations/dependencies?
e.g. `windowless_cgl_application` is connected to `cgl_context`, likewise for `egl`, `glx`, `wgl`

jgsogo replied:
I see the project generates a `egl/glx/wgl` objects library, and it only generates the actual library when `xxx_context` is True. So, we can have `xxx_context=False` and the `xxx_application=True` and it will work.

jgsogo commented on recipes/magnum/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,40 @@
+                if getattr(self.options["magnum"], opt_name):
+                    available.append(executable)
+            except:
+                pass
```
I've added the explicit exception, but still, I need to evaluate the option, not just check it exists.

madebr replied:
Indeed. My bad.

jgsogo commented:
@madebr @SSE4 do you think I should try to reproduce more accurately the logic encoded in the build-system? Right now I'm basically removing options that make the build fail and running some evident checks between options. Logic in the library's build-system files is much more complex... 

On one hand, I prefer to merge this recipe and the future will help: people building with other combinations of options, maybe new library versions with different logic,... I understand it is inconvenient to come across a build error while building, that it´s the role of the package manager to check these options before trying to build, but we all know that right now the logic is written by the author's library in the build-system files and we can only try to reproduce it in the recipes.

madebr commented:
I think it's fine. It's a hard exercise to check everything.
Also, a future magnum release can have a different cmake behavior.

prince-chrismc commented:
Fast bot 🙀 

I agree. If someone needs more options they can contribute back to fix them or improve availability... let's give something for now!
